### PR TITLE
Add more unit tests for transaction syncer

### PR DESCRIPTION
**Pull request by freehan**

### CODE REVIEW

File: pkg/neg/syncers/transaction.go [hunk 1]
```diff
@@ -61,21 +61,23 @@ type transactionSyncer struct {
...
 		endpointLister: endpointLister,
 		recorder:       recorder,
 		cloud:          cloud,
 		zoneGetter:     zoneGetter,
-		retry:          NewDelayRetryHandler(func() { syncer.Sync() }, NewExponentialBackendOffHandler(maxRetries, minRetryDelay, maxRetryDelay)),
 	}
-	syncer.SetSyncFunc(ts.sync)
+	// Syncer implements life cycle logic
+	syncer := newSyncer(negSyncerKey, networkEndpointGroupName, serviceLister, recorder, ts)
+	// TransactionSyncer needs syncer interface for internals
```
Comment (bowei) [inline pkg/neg/syncers/transaction.go]: [minor] isn't transactionSyncer lowercase

File: pkg/neg/syncers/transaction.go [hunk 2]
```diff
@@ -285,7 +287,7 @@ func (s *transactionSyncer) commitTransaction(err error, networkEndpointMap map[
 	for encodedEndpoint := range networkEndpointMap {
 		entry, ok := s.transactions.Get(encodedEndpoint)
 		if !ok {
-			glog.Errorf("Endpoint %q was not found in the transaction table.")
+			glog.Errorf("Endpoint %q was not found in the transaction table. This is unexpected", encodedEndpoint)
```
Comment (bowei) [inline pkg/neg/syncers/transaction.go]: "This is unexpected" probably not needed (you are already logging an Errorf

File: pkg/neg/syncers/transaction_test.go [hunk 3]
```diff
@@ -44,6 +49,293 @@ const (
 	testService   = "svc"
 )
 
+func TestTransactionSyncNetworkEndpoints(t *testing.T) {
+	fakeGCECloud := gce.FakeGCECloud(gce.DefaultTestClusterValues())
```
Comment (bowei) [inline pkg/neg/syncers/transaction_test.go]: t.Parallel() if possible

File: pkg/neg/syncers/transaction_test.go [hunk 4]
```diff
@@ -44,6 +49,293 @@ const (
...
+
+			if !endpoints.Equal(endpointSet) {
+				t.Errorf("For case %q, in zone %q, expect endpoints == %v, but got %v", tc.desc, zone, endpoints, endpointSet)
+			}
+		}
+	}
+}
+
+func TestCommitTransaction(t *testing.T) {
+	s, transactionSyncer := newTestTransactionSyncer(gce.FakeGCECloud(gce.DefaultTestClusterValues()))
```
Comment (bowei) [inline pkg/neg/syncers/transaction_test.go]: t.Parallel()

File: pkg/neg/syncers/transaction_test.go [hunk 5]
```diff
@@ -609,3 +882,53 @@ func generateEndpointSet(initialIp net.IP, num int, instance string, targetPort
...
+	return ret
+}
+
+type testSyncer struct {
+	*syncer
+	SyncCount int
+}
+
+func (s *testSyncer) Sync() bool {
+	s.SyncCount += 1
```
Comment (bowei) [inline pkg/neg/syncers/transaction_test.go]: ++

File: pkg/neg/types/mock.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+type NetworkEndpointEntry struct {
+	NetworkEndpoint *compute.NetworkEndpoint
+	Healths         []*compute.HealthStatusForNetworkEndpoint
+}
+
+type NetworkEndpointStore map[meta.Key][]NetworkEndpointEntry
+
+func MockNetworkEndpointAPIs(fakeGCE *gce.GCECloud) {
+	(fakeGCE.Compute().(*cloud.MockGCE)).MockBetaNetworkEndpointGroups.X = NetworkEndpointStore{}
```
Comment (bowei) [inline pkg/neg/types/mock.go]: probably shorter to

m := (fakeGCE.Compute().(*cloud.MockGCE)
m.MockBeta... = ...
...

**Comment by freehan** [comment]
Fixed the comments

**Pull request merged**